#244: <mission-complete-bar-pixel>

hanszhang00 opened the pull request:
Resolves #244

Fixes mission complete progress bar 1 pixel for 'previous mission' even though none has been completed. The screen should not be drawing either blue or black bar when there is no previous mission or other users' completed.

To test, run on local host and on the web page, select a neighborhood that shows 0% complete. Once a mission is complete in that neighborhood, look at the pop-up mission complete progress pixel bar, there should be only the blue bar drawn, and no other color.

Before:
![da8ddcdc-69d7-11e6-8d71-8956346b5396](https://user-images.githubusercontent.com/45343348/60555452-bf2d9e00-9cf1-11e9-8935-fcfaaf8baa3b.png)

After:
<img width="808" alt="Screen Shot 2019-07-02 at 4 49 23 PM" src="https://user-images.githubusercontent.com/45343348/60555461-c8b70600-9cf1-11e9-8dee-ea0b559c52c6.png">

misaugstad commented:
@hanszhang00 were you able to reproduce this bug yourself in the first place?

misaugstad commented:
@hanszhang00 actually could you explain how you fixed the bug?

hanszhang00 commented:
@misaugstad  i am able to reproduce the bug for a few times for the newberg database. however, it seems to be not consistent (sometimes there is 1 pixel blue or black bar drawn, while other times there is no problem). 

I made changes in ModalMissionComplete.js. For the update function in line 237, I found out that the missionDistance, userAuditedDistance, allAuditedDistance, as well as otherAuditedDistance are all in float values, and the float values will cause potential problems because the otherAuditedDistance is calculated by subtracting userAuditedDistance from allAuditedDistance, which won't guaranteed to be zero when it should be (inaccurate nature of floating value arithmetic). Thus what I did is I rounded all these variables to 0.1 accuracy (as required for the mission complete distance data accuracy), and I console logged it to make sure it is outputting the correct values.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionComplete.js:
```diff
@@ -240,12 +239,28 @@ function ModalMissionComplete (svl, missionContainer, missionModel, taskContaine
+        console.log("User has finished a particular mission...");
+
         var missionDistance = mission.getDistance("miles");
+        missionDistance = Math.round(missionDistance * 10) / 10;
```
I have two comments for these. The first is, like you said, it would be cleaner for the code to looks like
```
var missionDistance = mission.getDistance("miles").toFixed(1);
```

The second point is that you are rounding numbers that are then used in later calculations, which can cause the later numbers to more inaccurate. I think that what you want to do is round these numbers only when you actually use them. Like on develop right now, the `missionDistance` is not rounded until we actually use it in `_updateMissionProgressStatistics` function. So if the problem is that things should be rounded in the `modalMissionProgressBar.update`, just round them in that function and not earlier.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionComplete.js:
```diff
@@ -108,14 +108,13 @@ function ModalMissionComplete (svl, missionContainer, missionModel, taskContaine
         } else {
             // TODO can we require that we have a new mission before doing this?
             var nextMission = missionContainer.getCurrentMission();
-            _modalModel.triggerMissionCompleteClosed( { nextMission: nextMission } );
```
It looks like maybe you have some setting turned on in your editor that automatically updates the code style? Could you turn that off? Although I agree with most of the style changes, it distracts from the code you are actually trying to change.

hanszhang00 commented:
I have just made update.
I have deleted all the changes I made in the ModalMissionComplete.js file, and following what you suggested, fixed the rounding issue instead in the ModalMissionCompleteProgressBar.js.

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionCompleteProgressBar.js:
```diff
@@ -71,6 +71,11 @@ function ModalMissionCompleteProgressBar (uiModalMissionComplete) {
      * @private
      */
     this.update = function (missionDistanceRate, userAuditedDistanceRate, otherAuditedDistanceRate) {
+        // rounding rates to .1 accuracy to avoid floating value arithmetic error
```
As per [the style guide](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Style-Guide), could you change the comment to start with a capital letter and end with a period?

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionCompleteProgressBar.js:
```diff
@@ -71,6 +71,11 @@ function ModalMissionCompleteProgressBar (uiModalMissionComplete) {
      */
     this.update = function (missionDistanceRate, userAuditedDistanceRate, otherAuditedDistanceRate) {
+        // rounding rates to .1 accuracy to avoid floating value arithmetic error
+        missionDistanceRate = missionDistanceRate.toFixed(1);
```
You should avoid overwriting the values of input parameters. Instead, you should create new variables and (e.g., `roundedMissionDistanceRate`). This is just good practice, and in JavaScript it can have [a weird effect or two](https://spin.atomicobject.com/2011/04/10/javascript-don-t-reassign-your-function-arguments/).

hanszhang00 replied:
Style issues fixed.

misaugstad commented:
Ok I just tested and I am getting the error pictured below, and I am seeing no blue or green progress bar :( Are you getting the same thing?

![progress-bar-error](https://user-images.githubusercontent.com/6518824/60849183-89c7fb00-a19e-11e9-8a87-a25b8ac27453.png)

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionCompleteProgressBar.js:
```diff
@@ -71,25 +71,31 @@ function ModalMissionCompleteProgressBar (uiModalMissionComplete) {
     this.update = function (missionDistanceRate, userAuditedDistanceRate, otherAuditedDistanceRate) {
+
+        // Round the rates to 0.01 accuracy.
+        var roundedMissionDistanceRate = parseFloat(missionDistanceRate.toFixed(2));
```
This is looking a lot better! I think you actually want to go to 3 decimal places instead of 2. At 2 decimal places, the smallest amount we will show is 0.01, and 0.01 * 760 = 7.6. Which means that the smallest granularity we can go to is 7 pixels. But if you go to 3 places after the decimal, the smallest granularity is 1 pixel (0.001 * 760 = 0.76 ~ 1 pixel).

misaugstad commented on public/javascripts/SVLabel/src/SVLabel/modal/ModalMissionCompleteProgressBar.js:
```diff
@@ -71,25 +71,31 @@ function ModalMissionCompleteProgressBar (uiModalMissionComplete) {
             .duration(600)
-            .attr("width", missionDistanceRate * svgCoverageBarWidth);
-        horizontalBarMissionLabel.text(parseInt((userAuditedDistanceRate + missionDistanceRate) * 100, 10) + "%");
+            .attr("width", roundedMissionDistanceRate * svgCoverageBarWidth);
```
Just a quick idea I had: even if the mission the user did is not worth even 1 pixel, I think we should ensure that they still get a green bar. So I would change this line to
```
.attr("width", Math.max(1, roundedMissionDistanceRate * svgCoverageBarWidth));
```

hanszhang00 commented:
Changes done.
